src/lib/Imp.hs

```diff
@@ -269,178 +305,311 @@ toImpHof env (maybeDest, hof) = do
+    --     tileOffset <- toScalarAtom <$> iTile `imulI` tileLen
+    --     let tileAtom = Con $ IndexSliceVal idxTy tileIdxTy tileOffset
+    --     tileDest <- destSliceDim dest d tileOffset tileIdxTy
+    --     void $ translateBlock (env <> tb @> tileAtom) (Just tileDest, tBody)
+    --   emitLoop (binderNameHint sb) Fwd nEpilogue $ \iEpi -> do
+    --     i <- iEpi `iaddI` epilogueOff
+    --     idx <- intToIndex idxTy i
+    --     sDest <- destGetDim dest d idx
+    --     void $ translateBlock (env <> sb @> idx) (Just sDest, sBody)
+    --   destToAtom dest
     While (Lam (Abs _ (_, cond))) (Lam (Abs _ (_, body))) -> do
-      (condAtom, cond') <- scopedBlock $ translateBlock env (Nothing, cond)
-      (_, body') <- scopedBlock $ void $ translateBlock env (Nothing, body)
-      emitStatement $ IWhile (cond', Just $ fromScalarAtom condAtom) body'
+      cond' <- liftM snd $ scopedBlock $ do
+                 ans <- translateBlock env (Nothing, cond)
+                 return ((), [fromScalarAtom ans])
+      body' <- scopedVoidBlock $ void $ translateBlock env (Nothing, body)
+      emitStatement $ IWhile cond' body'
       return UnitVal
     RunReader r (BinaryFunVal _ ref _ body) -> do
       rDest <- alloc $ getType r
-      rVar  <- freshVar $ fromBind "ref" ref
       copyAtom rDest =<< impSubst env r
-      localRef rVar rDest $
-        translateBlock (env <> ref @> Var rVar) (maybeDest, body)
+      translateBlock (env <> ref @> rDest) (maybeDest, body)
     RunWriter (BinaryFunVal _ ref _ body) -> do
       (aDest, wDest) <- destPairUnpack <$> allocDest maybeDest resultTy
-      zeroDest $ wDest
-      wVar <- freshVar $ fromBind "ref" ref
-      void $ localRef wVar wDest $
-        translateBlock (env <> ref @> Var wVar) (Just aDest, body)
+      let RefTy _ wTy = getType ref
+      copyAtom wDest (zeroAt wTy)
+      void $ translateBlock (env <> ref @> wDest) (Just aDest, body)
       PairVal <$> destToAtom aDest <*> destToAtom wDest
     RunState s (BinaryFunVal _ ref _ body) -> do
       (aDest, sDest) <- destPairUnpack <$> allocDest maybeDest resultTy
       copyAtom sDest =<< impSubst env s
-      sVar <- freshVar $ fromBind "ref" ref
-      void $ localRef sVar sDest $
-        translateBlock (env <> ref @> Var sVar) (Just aDest, body)
+      void $ translateBlock (env <> ref @> sDest) (Just aDest, body)
       PairVal <$> destToAtom aDest <*> destToAtom sDest
     _ -> error $ "Invalid higher order function primitive: " ++ pprint hof
+
+abstractPtrLiterals :: Block -> ([IBinder], [LitVal], Block)
+abstractPtrLiterals block = flip evalState initState $ do
+  block' <- traverseLiterals block $ \val -> case val of
+    PtrLit ty ptr -> do
+      ptrName <- gets $ M.lookup (ty, ptr). fst
+      case ptrName of
+        Just v -> return $ Var $ v :> getType (Con $ Lit val)
+        Nothing -> do
+          (varMap, usedNames) <- get
+          let name = genFresh (Name AbstractedPtrName "ptr" 0) usedNames
```

Comment:
Couldn't we just extend the imp scope here? This would no longer be a pure function, but that shouldn't be an issue

Reply:
Yes, that would work too.